### PR TITLE
Dead code elimination

### PR DESCRIPTION
**Pull request by filipenevola**
This is a work in progress.

## Documentation

Support for custom flags with static replacement can be used by adding the following on your settings.json:
```
    "globalDefinitions": {
      "isAdminDashboard": false
    }
```
This will make Meteor.isAdminDashboard = false, and Meteor.isAdminDashboard be replaced with false when building, so a code snippet like:
```
if(Meteor.isAdminDashboard){
   console.log('huge string');
}
```
will be removed if you are building with "isAdminDashboard": false. This is useful when sharing the same code base for multiple projects.

## Tasks

- [x] 1. Use babel for replacing meteor internals like Meteor.isServer, Meteor.isClient, for better dead code elimination
- [x] 2. Expose it as an package, for setting custom build-time flags.
- [ ] 3. Gather imports/exports in a map for tree-shaking
- [ ] 4. Flag sideEffectsFree modules
- [ ] 5. Remove/don't include safe modules in the final bundle

### CODE REVIEW

**Comment by sebakerckhof** [comment]
Happy to give this a test on our apps if it's ready for review.

**Comment by filipenevola** [comment]
@sebakerckhof dead code elimination is working already. See the instructions in the description if you also want to have custom dead code elimination applied. Please provide feedback.

Now @renanccastro is proceeding with tree-shaking itself, starting by node_modules, we are going to keep this PR updated.

**Comment by sebakerckhof** [comment]
@filipenevola Maybe it could also set Meteor.isModern to true for modern web bundle and false for legacy web bundle.

File: packages/minifier-js/minifier.js [hunk 1]
```diff
@@ -1,35 +1,78 @@
 var terser;
 
-meteorJsMinify = function (source) {
+const getGlobalDefsOptions = ({ arch }) => ({
+  "Meteor.isServer": false,
+  "Meteor.isTest": false,
+  "Meteor.isDevelopment": false,
+  "Meteor.isClient": true,
+  "Meteor.isProduction": true,
+  "Meteor.isCordova": arch === 'web.cordova',
```
Comment (sebakerckhof) [inline packages/minifier-js/minifier.js:9]: What about adding:
"Meteor.isModern": arch !== 'web.browser.legacy'

That's assuming cordova is also modern.

**Comment by Fen747** [comment]
Hi !

First, thank you for your work on this :)

As I'm currently struggling with `formik-antd` importing all of `antd`, and unable to find a working solution with `babel`, I'd like to ask if this PR will eventually be able to prune such overkill imports that are made directly from within a npm module ?

This will be a game changer for Meteor when it comes out ;)

Cheers !

**Comment by renanccastro** [comment]
> Hi !
> 
> First, thank you for your work on this :)
> 
> As I'm currently struggling with `formik-antd` importing all of `antd`, and unable to find a working solution with `babel`, I'd like to ask if this PR will eventually be able to prune such overkill imports that are made directly from within a npm module ?
> 
> This will be a game changer for Meteor when it comes out ;)
> 
> Cheers !

Hello! Yes, this feature I'm working on: https://github.com/meteor/meteor/pull/11164 will do what you said, and will also evolve this work ;)

for more details, please check the pr I've linked.

I will close this PR for now, as the development will keep on [this](https://github.com/meteor/meteor/pull/11164) one.

**Pull request closed without merging**